SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,180 @@
+///    that there is a balancing release for the new "copy" partial apply. But
+///    since the original partial apply no longer will have that corresponding
+///    -1, we need to insert a release for the old partial apply. We do this
+///    right after the old call site where the original partial apply was
+///    called. This ensures we do not shrink the lifetime of the old partial
+///    apply. In the case where the old partial_apply was passed in at +0, we
+///    know that the old partial_apply does not need to have any ref count
+///    adjustments. On the other hand, the new "copy" partial apply in the
+///    specialized function now needs to be balanced lest we leak. Thus we
+///    insert a release right before any exit from the function. This ensures
+///    that the release occurs in the epilog after any retains associated with
+///    @owned return values.
+///
+/// 3. In !useLoweredAddresses mode, we do not support specialization of closures
+///    with arguments passed using any indirect calling conventions besides
+///    @inout and @inout_aliasable.  This is a temporary limitation that goes
+///    away with sil-opaque-values.
+
+import SIL
+
+private let verbose = true
+
+private func log(_ message: @autoclosure () -> String) {
+  if verbose {
+    print("### \(message())")
+  }
+}
+
+// =========== Entry point into the pass =========== //
+let autodiffClosureSpecialization = FunctionPass(name: "autodiff-closure-specialize") {
+    (function: Function, context: FunctionPassContext) in
+
+    // 1\ Don't optimize functions that are marked with the opt.never attribute
+    // 2\ If F is an external declaration, there is nothing to specialize
+    // 3\ Specialize
+    // 4\ Eliminate dead closures
+    // 5\ Invalidate everything using `invalidateAnalysis` since we delete calls 
+    //    as well as add new calls and branches
+    
+    // var stackAllocDemangler = StackAllocatedDemangler().asDemangler()
+    // let demangler = Demangler()
+    // demangler.providePreallocatedMemory(parent: &stackAllocDemangler)
+    // demangler.demangleSymbol(mangledName: function.name.string)
+}
+
+// =========== AutodiffClosureSpecializer ========== //
+struct AutodiffClosureSpecializer: ClosureSpecializer {
+    typealias PropagatedClosure = SingleValueInstruction
+
+    func gatherCallSites(_ caller: Function) -> Stack<CallSiteDescriptor> {
+        fatalError("Not implemented")
+    }
+
+    func specialize(_ caller: Function) -> Stack<PropagatedClosure> {
+        fatalError("Not implemented")
+    }
+}
+
+// ===================== Types ===================== //
+
+/// Represents all the information required to
+/// represent a closure in isolation, i.e., outside of
+/// a callsite context where the closure may be getting
+/// passed as an argument.
+///
+/// Composed with other information inside a `ClosureArgDescriptor`
+/// to represent a closure as an argument at a callsite.
+class ClosureInfo {
```

Comment:
Why does this need to be a class?

Reply:
I wanted to be able to use a common `lifetimeFrontier` value if this closure is present at multiple callsites or multiple times at one callsite.